core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -580,24 +587,37 @@ private void closeAndNotify(DiscoveryNode node, NodeChannels nodeChannels, Strin
     /**
      * Disconnects from a node if a channel is found as part of that nodes channels.
      */
-    protected final void disconnectFromNodeChannel(final Channel channel, final Exception failure) {
+    protected final void disconnectFromNodeChannel(final Channel channel, final String reason) {
         threadPool.generic().execute(() -> {
             try {
-                try {
+                if (isOpen(channel)) {
                     closeChannels(Collections.singletonList(channel));
-                } finally {
-                    for (DiscoveryNode node : connectedNodes.keySet()) {
-                        if (disconnectFromNode(node, channel, ExceptionsHelper.detailedMessage(failure))) {
+                }
+            } catch (IOException e) {
+                logger.warn("failed to close channel", e);
+            } finally {
+                outer:
+                {
+                    for (Map.Entry<DiscoveryNode, NodeChannels> entry : connectedNodes.entrySet()) {
+                        if (disconnectFromNode(entry.getKey(), channel, reason)) {
                             // if we managed to find this channel and disconnect from it, then break, no need to check on
                             // the rest of the nodes
+                            // Removing it here from open connections is paranoia since the close handler should deal with this.
+                            openConnections.remove(entry.getValue());
+                            // we can only be connected and published to a single node with one connection. So if disconnectFromNode
+                            // returns true we can safely break out from here since we cleaned up everything needed
+                            break outer;
+                        }
+                    }
+                    // now if we haven't found the right connection in the connected nodes we have to go through the open connections
+                    // it might be that the channel belongs to a connection that is not published
+                    for (NodeChannels channels : openConnections) {
```

Comment:
@s1monw and I discussed this and apparently transportServiceAdapter spawns threads and any point of synchronization is moot and only a source of trouble (blocking on netty is super tricky)